kb/how-to/rewrite-branch-for-coding-style.md

```diff
@@ -0,0 +1,163 @@
+         /
+D---E---F---G---H target
+            ^
+            |
+            "Switch to the new code style"
+```
+
+In the second phase, the migration successively constructs a new branch with restyled commits. First the migration restyles the content of A' to obtain A'':
+
+```
+            A'---B'---C' mywork
+           /
+          /   A'' new/mywork
+         /   /
+D---E---F---G---H target
+            ^
+            |
+            "Switch to the new code style"
+```
+
+The process continues with B' and so on:
+
+```
+            A'---B'---C' mywork
+           /
+          /   A''---B'' new/mywork
+         /   /
+D---E---F---G---H target
+            ^
+            |
+            "Switch to the new code style"
+```
+
+Once the migration reaches the end of the branch, it renames the new branch to the original name:
+
+```
+            A'---B'---C' old/mywork
+           /
+          /   A''---B''---C'' mywork
+         /   /
+D---E---F---G---H target
+            ^
+            |
+            "Switch to the new code style"
+```
+
+The branch `mywork` can now be pushed to a pull request.
+
+## When to migrate
+
+If you have not yet made a pull request, you can migrate your work at any time.
+
+If you have opened a pull request but review has not yet started on it, please migrate your branch to the new code style before review starts.
+
+If there is an ongoing review of a pull request, please communicate with the reviewer(s) to decide on a good time. Rebasing a branch during review is disruptive, because comments end up detached from the code they're about, so it's generally best not to do it. Typically the best time to migrate the branch is after the reviewer(s) have approved.
+
+## How to migrate
+
+### Requirements
+
+Please have the following tools and files available:
+
+* Git 2.17.0 or newer.
+* Uncrustify 0.75.1. Please note that older or newer versions are likely not to give the desired result.
+* Python 3.6 or newer.
+* A Git worktree containing the branch of Mbed TLS you want to migrate.
+* A local copy of [`scripts/code_style_rewrite_branch.py`](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/code_style_rewrite_branch.py) (preferably the most recent version from `development`, since it may have received bug fixes).
+
+### Simple migration
+
+Prerequisites:
+
+* You have the tools and files listed under [“Requirements”](#requirements) above.
+* You have a Git branch `mywork` which branched from `development` before the code style switch. (For work on Mbed TLS 2.28, replace `development` by `mbedtls-2.28` throughout these instructions.)
+* The current directory is a Git working directory with `mywork` checked out, and with a remote called `upstream` pointing to [the official Mbed TLS repository](https://github.com/Mbed-TLS/mbedtls).
+
+Run the following command in a terminal, from the Git working directory with `mywork` checked out:
+
+```sh
+python3 /path/to/code_style_rewrite_branch.py mywork upstream/development
```

Comment:
So, we expect the script will be able to locate the restyling commit from the target branch?

Reply:
Yes: the commit can be discovered as `upstream/development^{/Switch to the new code style}` (except we need to protect against another commit including this string for some reason). Or we can even autodetect the branch between development and 2.28, and hard-code their respective commit IDs. To be updated once I have the new script working.